File: Front/.Net Maui/BetFriends.Domain/Features/Register/RegisterHandler.cs
using BetFriends.Domain.Abstractions;
using MediatR;

namespace BetFriends.Domain.Features.Register;

public sealed class RegisterHandler(IRegisterOutputPort presenter,
                                    IUserGateway userGateway,
                                    IIdGenerator idGenerator,
                                    IHashPassword hashPassword) : IRequestHandler<RegisterRequest>
{
    private IRegisterOutputPort presenter = presenter;

    public async Task Handle(RegisterRequest request, CancellationToken cancellationToken)
    {
        if (IsNotValidRequest(request))
            return;

        if (IsPasswordsAreNotEqual(request))
        {
            presenter.PasswordsNotEqual();
            return;
        }

        var password = hashPassword.Hash(request.Password);
        var user = new User(idGenerator.Generate(), request.Username, request.Email, password);
        await userGateway.SaveAsync(user);
        presenter.Present(user.Id);
    }

    private static bool IsPasswordsAreNotEqual(RegisterRequest request)
        => request.Password != request.ConfirmPassword;

    private bool IsNotValidRequest(RegisterRequest request)
    {
        if (string.IsNullOrEmpty(request.Email)
                    || string.IsNullOrEmpty(request.Username)
                    || string.IsNullOrEmpty(request.Password)
                    || string.IsNullOrEmpty(request.ConfirmPassword))
        {
            presenter.FieldEmpty();
            return true;
        }
        return false;
    }
}

public interface IRegisterOutputPort
{
    void FieldEmpty();
    void PasswordsNotEqual();
    void Present(string id);
}

File: Front/TS/src/domain/features/RegisterHandler.ts
import { IIdGenerator } from "../abstractions/IIdGenerator";
import { IUserGateway } from "../abstractions/IUserGateway";
import { IHashService } from '../abstractions/IHashService';

export class RegisterHandler {

    constructor(private readonly userGateway: IUserGateway,
        private readonly outputPort: IRegisterOutputPort,
        private readonly idGenerator: IIdGenerator,
        private readonly passwordHasher: IHashService){}

        
    async Handle(request: IRegisterRequest): Promise<void> {
        if(request.confirmPassword.length == 0 ||
            request.username.length == 0 ||
            request.email.length == 0 ||
            request.password.length == 0) {
            this.outputPort.FieldIsEmpty()
            return;
        }

        if(request.confirmPassword != request.password) {
            this.outputPort.PasswordsAreDifferent()
            return;
        }
        const userId = this.idGenerator.Generate()
        const password = this.passwordHasher.Hash(request.password)
        const user = {
            id: userId,
            username: request.username,
            email: request.email,
            password: password
        }
        await this.userGateway.Register(user)
        this.outputPort.UserRegistered(userId)
    }
}

export interface IRegisterRequest {
    username: string;
    email: string;
    password: string;
    confirmPassword: string;
}

export interface IRegisterOutputPort {
    PasswordsAreDifferent(): void;
    FieldIsEmpty(): void;
    UserRegistered(userId: string): void;

}